test/mqtt5_canary.py

```diff
@@ -0,0 +1,282 @@
+        )
+
+        if random.getrandbits(1):
+            publish_packet.user_properties = self.user_properties
+
+        try:
+            self.client.publish(publish_packet=publish_packet)
+            self.canary_core.stat_publishes_succeeded += 1
+        except BaseException:
+            pass
+
+    def subscribe(self, topic_filter: str = None, qos: int = 1):
+        self.canary_core.stat_subscribes_attempted += 1
+        if topic_filter is None:
+            topic_filter = str(time.time()) + self.client_id
+
+        self.canary_core.subscriptions.append(topic_filter)
+
+        subscribe_packet = mqtt5.SubscribePacket(subscriptions=[mqtt5.Subscription(topic_filter=topic_filter, qos=qos)])
+
+        try:
+            self.client.subscribe(subscribe_packet=subscribe_packet)
+            self.canary_core.stat_subscribes_succeeded += 1
+        except BaseException:
+            pass
+
+    def unsubscribe(self):
+        if len(self.canary_core.subscriptions) < 1:
+            return
+
+        self.canary_core.stat_unsubscribes_attempted += 1
+        unsubscribe_packet = mqtt5.UnsubscribePacket(topic_filters=[self.canary_core.subscriptions.pop()])
+
+        try:
+            self.client.unsubscribe(unsubscribe_packet=unsubscribe_packet)
+            self.canary_core.stat_unsubscribes_succeeded += 1
+        except BaseException:
+            pass
+
+    def print_stats(self):
+        print(f"""\n
+Client Stats:
+    publishes_received:     {self.canary_core.stat_publishes_received}
+
+    subscribes_attempted:   {self.canary_core.stat_subscribes_attempted}
+    subscribes_succeeded:   {self.canary_core.stat_subscribes_succeeded}
+
+    unsubscribes_attempted: {self.canary_core.stat_unsubscribes_attempted}
+    unsubscribes_succeeded: {self.canary_core.stat_unsubscribes_succeeded}
+
+    publishes_attempted:    {self.canary_core.stat_publishes_attempted}
+    publishes_succeeded:    {self.canary_core.stat_publishes_succeeded}
+
+    total_starts:           {self.canary_core.stat_total_starts}
+    total_stops:            {self.canary_core.stat_total_stops}
+
+    total operations:       {self.canary_core.stat_total_operations}""", file=sys.stdout)
+
+
+if __name__ == '__main__':
+    client = CanaryClient()
+    time_end = time.time() + float(seconds)
+    tpsdelay = 1.0 / float(tps)
+    time_next_operation = time.time()
+
+    print(f"""\n
+    Canary running for {seconds} seconds
+    TPS: {tps}
+    Clients: {client_count}
+    Threads: {threads}
+    """, file=sys.stdout)
+
+    clients = []
+    for i in range(int(client_count)):
+        clients.append(CanaryClient())
+
+    for client in clients:
+        client.initial_start()
+
+    for client in clients:
+        client.subscribe()
+
+    while time.time() < time_end:
+        if time.time() >= time_next_operation:
+            time_next_operation += tpsdelay
+            for client in clients:
+                client.random_operation()
```

Comment:
A minor issue, the behavior for random operation here is different from the one in aws-c-mqtt. In aws-c-mqtt, it runs 1 operation/tpsdelay, while in python it runs 1 operation/client/tpsdelay. 
I think either behavior works, just need to note that it actually runs N*tps operations per seconds in total.

Reply:
Ah! Great catch! Fixed to run the correct TPS.